src/dataflow/src/source/mod.rs

```diff
@@ -861,6 +867,117 @@ enum MessageProcessing {
     YieldedWithDelay,
 }
 
+#[derive(Clone)]
+/// A thread-safe shareable timestamp provider that allows consumers to pause ticking of the clock
+/// while they are in the middle of producing data for a particular timestamp.
+pub struct Timestamper {
```

Comment:
The Timestamper is nice but the current API is a bit of a footgun:

1) The act of releasing a lease has some rust magic that's unclear. For example, in the cockroach cdc code I played with, I needed to do `time_lease = None; time_lease = Some(timestamper.lease().await);`. Without the explicit call to `time_lease = None` it'd stall the process (and some lint complains about the unused double assign). I don't know rust well enough to know what's going on, but it was confusing to me, and won't pass our CI due to the lint for time sources that don't have explicit begin/end txn events like cockroach cdc.
2) There's no enforcement that the time sent over the stream has come from the timestamper. The Event::Progress messages make assumptions about the times being produced by the stream, but the stream could emit any times it wants.

I don't think it's essential that these get fixed here. But perhaps these issues should be documented so future users of Timestamper don't do wrong things.

Reply:
> The act of releasing a lease has some rust magic that's unclear.

This isn't really rust magic, it's just the RAII pattern which we use in a lot of places. Timely's capabilities are one example that do the exact same thing. Take a look at this very similar code from the existing source implementation: https://github.com/MaterializeInc/materialize/blob/82bf406b2026bfe03eed9f048d9ba083ecddcc7c/src/dataflow/src/source/util.rs#L80-L81

> time sources that don't have explicit begin/end txn events like cockroach cdc.

I'm not familiar with cockroach cdc, but if a source doesn't have a notion of transactions in the first place then it's also unlikely that the implementer will be holding onto timestamps across rows. There is one potential issue though if someone writes code like this:

```rust
let time = timestamper.lease().await;
let row = my_stream.next().await; // The lease will be locked until the next row arrives
```

> There's no enforcement that the time sent over the stream has come from the timestamper.

That's a good point. I'll try to encode this invariant in the types

Reply:
Sure, that pattern makes sense. The exact problem I had was that:

```
time_lease = Some(timestamper.lease().await);
```

stalled the process but

```
time_lease = None;
time_lease = Some(timestamper.lease().await);
```

did not. Even though `time_lease` was manually assigned twice, and thus the first assignment shouldn't matter, it seemed to. (I suppose it's possible I was observing some other bug, but I recall having to add an explicit `= None` in order to make it progress.)

Cockroach cdc emits "resolved timestamp" data regularly which is what I used as a signal to bump the timestamp.